// @remixproject/plugin-ws/src/lib/ws.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClient = exports.WebsocketConnector = void 0;
const plugin_utils_1 = require("@remixproject/plugin-utils");
const plugin_1 = require("@remixproject/plugin");
/**
 * This Websocket connector works with the library `ws`
 */
class WebsocketConnector {
    constructor(websocket) {
        this.websocket = websocket;
    }
    setClient(client) {
        this.client = client;
    }
    /** Send a message to the engine */
    send(message) {
        this.websocket.send(JSON.stringify(message));
    }
    /** Get message from the engine */
    on(cb) {
        this.websocket.on('message', (event) => {
            try {
                const parsedEvent = JSON.parse(event);
                if (!plugin_1.isHandshake(parsedEvent)) {
                    if (parsedEvent.action &&
                        (parsedEvent.action === 'request' || parsedEvent.action === 'call')) {
                        const path = parsedEvent.requestInfo && parsedEvent.requestInfo.path;
                        const method = plugin_utils_1.getMethodPath(parsedEvent.key, path);
                        if (this.client.methods && !this.client.methods.includes(method)) {
                            throw new Error(`${method} is not in the list of allowed methods.`);
                        }
                    }
                }
                cb(JSON.parse(event));
            }
            catch (e) {
                console.error(e);
            }
        });
    }
}
exports.WebsocketConnector = WebsocketConnector;
/**
 * Connect a Websocket plugin client to a web engine
 * @param client An optional websocket plugin client to connect to the engine.
 *
 * ---------
 * @example
 * ```typescript
 * const wss = new WebSocket.Server({ port: 8080 });
 * wss.on('connection', (ws) => {
 *  const client = createClient(ws)
 * })
 * ```
 * ---------
 * @example
 * ```typescript
 * class MyPlugin extends PluginClient {
 *  methods = ['hello']
 *  hello() {
 *   console.log('Hello World')
 *  }
 * }
 * const wss = new WebSocket.Server({ port: 8080 });
 * wss.on('connection', (ws) => {
 *  const client = createClient(ws, new MyPlugin())
 * })
 * ```
 */
exports.createClient = (websocket, client = new plugin_1.PluginClient()) => {
    const c = client;
    const websocketConnector = new WebsocketConnector(websocket);
    plugin_1.connectClient(websocketConnector, c);
    plugin_1.applyApi(c);
    websocketConnector.setClient(c);
    return c;
};
